### PR TITLE
Initial Atomic KMS platform

### PR DESCRIPTION
**Pull request by RAOF**
This is a minimal-changes copy of the display half of the `gbm-kms` platform, switched over to all Atomic KMS API.

It's parallel installable with the existing `gbm-kms` platform, which will still be preferentially used on devices with atomic support.

Currently, over the existing `gbm-kms` display platform it only has improved gamma support.
Immediate missing features:
* Hardware cursor support
* Querying list of supported pixel formats

### CODE REVIEW

File: src/platforms/atomic-kms/server/kms/quirks.cpp [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+{
+    return impl->should_skip(device);
+}
+
+void mga::Quirks::add_quirks_option(boost::program_options::options_description& config)
+{
+    config.add_options()
+        (quirks_option_name,
+         boost::program_options::value<std::vector<std::string>>(),
+         "[platform-specific] Driver quirks to apply (may be specified multiple times; multiple quirks are combined)");
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/quirks.cpp:185]: This same option is added by `gbm-kms` - and that spells trouble if both platforms are available. (Similar to #885)

**Comment by Saviq** [comment]
Hmm with dual outputs I get:

    < - ERROR - > atomic-kms: Failed to set CRTC: No space left on device (28)

And an occasional SIGABRT after.

Yes. I. Have. Disk. Space.

**Comment by Saviq** [comment]
@tarek-y-ismail's ploughed through drm debug logs and found:

```
[drm:drm_atomic_plane_check] [PLANE:32:plane 1A] invalid source coordinates 2256.000000x1504.000000+0.000000+0.000000 (fb 1920x1200)
[drm:drm_atomic_check_only] [PLANE:32:plane 1A] atomic core check failed
```

This suggests there's a problem with cloning, as it tries to map the smaller buffer onto a bigger display.

**Comment by tarek-y-ismail** [comment]
> Hmm with dual outputs I get:

By "dual outputs", do you mean buitlin-monitor + external monitor, or two external monitors?
I don't get this error, but I also don't get external output (no parameters passed except `--platform-display-libs=mir:atomic-kms --platform-rendering-libs=mir:gbm-kms`)

**Comment by Saviq** [comment]
> By "dual outputs", do you mean builtin-monitor + external monitor, or two external monitors?

Bulting + external.

> I don't get this error, but I also don't get external output.

As discussed, to get output through your Nvidia GPU, you'll need `,mir:eglstream-kms` added to your display plugins.

**Comment by tarek-y-ismail** [comment]
Just tried it, both monitors work fine. I assume this is because the second output is not using atomic-kms and thus not triggering the bug

**Comment by Saviq** [comment]
> Just tried it, both monitors work fine. I assume this is because the second output is not using atomic-kms and thus not triggering the bug

Yup, that could well be it - they're not sharing the platform like in my case.

File: src/platforms/atomic-kms/server/kms/quirks.cpp (hunk 1 above)
Comment (tarek-y-ismail) [reply src/platforms/atomic-kms/server/kms/quirks.cpp:185]: Is this resolved by #3590?

**Comment by Saviq** [comment]
Trying heterogeneous refresh rates I got:

```
[2024-10-09 18:19:28.744101] <information> mirserver: Initial display configuration:
[2024-10-09 18:19:28.744134] <information> mirserver: * Output 1: eDP connected, used
[2024-10-09 18:19:28.744156] <information> mirserver: . |_ EDID manufacturer: BOE
[2024-10-09 18:19:28.744172] <information> mirserver: . |_ EDID product code: 2399
[2024-10-09 18:19:28.744190] <information> mirserver: . |_ Physical size 13.3" 280x190mm
[2024-10-09 18:19:28.744203] <information> mirserver: . |_ Power is on
[2024-10-09 18:19:28.744216] <information> mirserver: . |_ Current mode 2256x1504 59.99Hz
[2024-10-09 18:19:28.744228] <information> mirserver: . |_ Preferred mode 2256x1504 59.99Hz
[2024-10-09 18:19:28.744240] <information> mirserver: . |_ Orientation normal
[2024-10-09 18:19:28.744255] <information> mirserver: . |_ Logical size 2256x1504
[2024-10-09 18:19:28.744268] <information> mirserver: . |_ Logical position +0+0
[2024-10-09 18:19:28.744280] <information> mirserver: . |_ Scaling factor: 1.00
[2024-10-09 18:19:28.744292] <information> mirserver: * Output 2: DisplayPort connected, used
[2024-10-09 18:19:28.744305] <information> mirserver: . |_ EDID monitor name: DELL U2413
[2024-10-09 18:19:28.744326] <information> mirserver: . |_ EDID manufacturer: DEL
[2024-10-09 18:19:28.744338] <information> mirserver: . |_ EDID product code: 61512
[2024-10-09 18:19:28.744351] <information> mirserver: . |_ Physical size 24.0" 520x320mm
[2024-10-09 18:19:28.744364] <information> mirserver: . |_ Power is on
[2024-10-09 18:19:28.744377] <information> mirserver: . |_ Current mode 1920x1080 23.97Hz
[2024-10-09 18:19:28.744403] <information> mirserver: . |_ Preferred mode 1920x1200 59.95Hz
[2024-10-09 18:19:28.744416] <information> mirserver: . |_ Orientation normal
[2024-10-09 18:19:28.744433] <information> mirserver: . |_ Logical size 1920x1080
[2024-10-09 18:19:28.744445] <information> mirserver: . |_ Logical position +2256+0
[2024-10-09 18:19:28.744457] <information> mirserver: . |_ Scaling factor: 1.00
[2024-10-09 18:19:28.744474] <information> mirserver: * Output 3: DisplayPort disconnected
[2024-10-09 18:19:28.744487] <information> mirserver: * Output 4: DisplayPort disconnected
[2024-10-09 18:19:28.744503] <information> mirserver: * Output 5: DisplayPort disconnected
[2024-10-09 18:19:28.744747] < - debug - > miral: Configuring pointer: 'basic-window-manager'
[2024-10-09 18:19:28.744814] <information> input-hub: Device configuration: basic-window-manager, capabilities={pointer}
[2024-10-09 18:19:28.761780] < - ERROR - > atomic-kms: Failed to set CRTC: Invalid argument (22)
!!! Fatal signal received. Attempting cleanup, but deadlock may occur
Mir fatal error: Unsupported attempt to continue after a fatal signal: SIGSEGV
!!! Fatal signal received. Attempting cleanup, but deadlock may occur
Mir fatal error: Unsupported attempt to continue after a fatal signal: SIGABRT
```
[drm.log](https://github.com/user-attachments/files/17311521/drm.log)

**Comment by RAOF** [comment]
Oh! It's not heterogeneous *refresh* rates, it's heterogeneous *resolutions*!

**Comment by Saviq** [comment]
So we're clear on the issue I'm seeing still:

https://github.com/user-attachments/assets/42f0560b-0a25-4acf-ad60-e5c59b1eed30

**Comment by RAOF** [comment]
AHA! It took writing a script to try and reproduce this, but I now can.

What's happening here is that the final frame submitted by the terminal is racing the modeset sequence.

When it stops updating, what's happening is the terminal is waiting on a frame event from its last-buffer-before-modeset before submitting any new frames. Mir is not sending a frame event until *something else* triggers a composition.

So this is a problem somewhere in our buffer queuing, I think.

File: src/platforms/atomic-kms/server/display_helpers.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+#include "mir/fd.h"
+
+#include <cstddef>
+#include <memory>
+#include <vector>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic warning "-Wall"
+#include <gbm.h>
+#pragma GCC diagnostic pop
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/display_helpers.h:30]: I'm curious: most places we include `<gbm.h>` we don't disable warnings. Why is this occasion different?

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+using LockedFrontBuffer = std::unique_ptr<gbm_bo, std::function<void(gbm_bo*)>>;
+
+class GBMBoFramebuffer : public mg::FBHandle
+{
+public:
+    static auto framebuffer_for_frontbuffer(mir::Fd const& drm_fd, LockedFrontBuffer bo) -> std::unique_ptr<GBMBoFramebuffer>
+    {
+        if (auto cached_fb = static_cast<std::shared_ptr<uint32_t const>*>(gbm_bo_get_user_data(bo.get())))
+        {
+            return std::unique_ptr<GBMBoFramebuffer>{new GBMBoFramebuffer{std::move(bo), *cached_fb}};
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/gbm_display_allocator.cpp:58]: AFAICS, the only reason for using raw `new` is that the constructor is private. But, as the class is local to this translation unit, and the member data is deep const, that seems a bad reason.

```suggestion
            return std::make_unique<GBMBoFramebuffer>(std::move(bo), *cached_fb);
```

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        auto fb_id = new std::shared_ptr<uint32_t>{
+            new uint32_t{0},
+            [drm_fd](uint32_t* fb_id)
+            {
+                if (*fb_id)
+                {
+                    drmModeRmFB(drm_fd, *fb_id);
+                }
+                delete fb_id;
+            }};
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/gbm_display_allocator.cpp:70]: Do we _really_ want to allocate a `shared_ptr<>` on the stack? And ultimately leak it (preventing the deleter ever being used)?

```suggestion
        auto fb_id = std::shared_ptr<uint32_t>{
            new uint32_t{0},
            [drm_fd](uint32_t* fb_id)
            {
                if (*fb_id)
                {
                    drmModeRmFB(drm_fd, *fb_id);
                }
                delete fb_id;
            }};
```

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+
+        /* Create a KMS FB object with the gbm_bo attached to it. */
+        auto ret = drmModeAddFB2(drm_fd, width, height, format,
+                                 handles, strides, offsets, fb_id->get(), 0);
+        if (ret)
+            return nullptr;
+
+        gbm_bo_set_user_data(bo.get(), fb_id, [](gbm_bo*, void* fb_ptr) { delete static_cast<std::shared_ptr<uint32_t const>*>(fb_ptr); });
+
+        return std::unique_ptr<GBMBoFramebuffer>{new GBMBoFramebuffer{std::move(bo), *fb_id}};
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/gbm_display_allocator.cpp:88]:
```suggestion
        return std::make_unique<GBMBoFramebuffer>(std::move(bo), std::move(fb_id));
```

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+auto create_gbm_surface(gbm_device* gbm, geom::Size size, mg::DRMFormat format, std::span<uint64_t> modifiers)
+    -> std::shared_ptr<gbm_surface>
+{
+    auto const surface =
+        [&]()
+        {
+            if (modifiers.empty())
+            {
+                // If we have no no modifiers don't use the with-modifiers creation path.
+                auto foo = GBM_BO_USE_RENDERING | GBM_BO_USE_SCANOUT;
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/gbm_display_allocator.cpp:125]:
```suggestion
                auto constexpr foo = GBM_BO_USE_RENDERING | GBM_BO_USE_SCANOUT;
```

File: src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp [hunk 7]
```diff
@@ -0,0 +1,831 @@
...
+        info1->clock == info2->clock &&
+        info1->hdisplay == info2->hdisplay &&
+        info1->hsync_start == info2->hsync_start &&
+        info1->hsync_end == info2->hsync_end &&
+        info1->htotal == info2->htotal &&
+        info1->hskew == info2->hskew &&
+        info1->vdisplay == info2->vdisplay &&
+        info1->vsync_start == info2->vsync_start &&
+        info1->vsync_end == info2->vsync_end &&
+        info1->vtotal == info2->vtotal;
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp:58]:
```suggestion
        std::tie(info1->clock, info1->hdisplay, ...) ==
        std::tie(info2->clock, info2->hdisplay, ...);
```

File: src/platforms/atomic-kms/server/kms/display.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_GRAPHICS_GBM_DISPLAY_H_
+#define MIR_GRAPHICS_GBM_DISPLAY_H_
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/display.h:18]:
```suggestion
#ifndef MIR_GRAPHICS_ATOMIC_GBM_DISPLAY_H_
#define MIR_GRAPHICS_ATOMIC_GBM_DISPLAY_H_
```

File: src/platforms/atomic-kms/server/kms/egl_helper.cpp [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+ */
+
+#include "egl_helper.h"
+#include "mir/graphics/gl_config.h"
+#include "mir/graphics/egl_error.h"
+#include <boost/exception/errinfo_errno.hpp>
+#include <boost/throw_exception.hpp>
+#include <gbm.h>
+
+#define MIR_LOG_COMPONENT "EGL"
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/egl_helper.cpp:24]: This doesn't distinguish from gbm

File: src/platforms/atomic-kms/server/kms/egl_helper.cpp [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+
+    if (eglInitialize(dpy, &major, &minor) == EGL_FALSE)
+        BOOST_THROW_EXCEPTION(mg::egl_error("Failed to initialize EGL display"));
+
+    if ((major < minimum_major_version) ||
+        (major == minimum_major_version && minor < minimum_minor_version))
+    {
+        BOOST_THROW_EXCEPTION(
+            boost::enable_error_info(std::runtime_error("Incompatible EGL version")));
+        // TODO: Insert egl version major and minor into exception
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/egl_helper.cpp:79]: Let's resolve the TODO before landing

File: src/platforms/atomic-kms/server/kms/egl_helper.cpp [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+void mgmh::EGLHelper::report_egl_configuration(std::function<void(EGLDisplay, EGLConfig)> f)
+{
+    f(egl_display, egl_config);
+}
+
+mgmh::EGLHelper::NoMatchingEGLConfig::NoMatchingEGLConfig(uint32_t /*format*/)
+    : std::runtime_error("Failed to find matching EGL config")
+{
+    // TODO: Include the format string; need to extract from linux_dmabuf.cpp
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/egl_helper.cpp:285]: Let's resolve the TODO

File: src/platforms/atomic-kms/server/kms/egl_helper.h [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef MIR_GRAPHICS_GBM_EGL_HELPER_H_
+#define MIR_GRAPHICS_GBM_EGL_HELPER_H_
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/egl_helper.h:18]:
```suggestion
#ifndef MIR_GRAPHICS_ATOMIC_EGL_HELPER_H_
#define MIR_GRAPHICS_ATOMIC_EGL_HELPER_H_
```

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
A bunch of surface cruft that needs cleaning away. At least one more review pass needed

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp (hunk 4 above)
Comment (AlanGriffiths) [reply src/platforms/atomic-kms/server/gbm_display_allocator.cpp:70]: On closer examination the code was copied from [elsewhere](https://github.com/canonical/mir/pull/3644) and we only sometimes leak the shared_ptr.

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+auto mga::GBMDisplayAllocator::supported_formats() const -> std::vector<DRMFormat>
+{
+    // TODO: Pull out of KMS plane info
+    return { DRMFormat{DRM_FORMAT_XRGB8888}, DRMFormat{DRM_FORMAT_ARGB8888}};
+}
+
+auto mga::GBMDisplayAllocator::modifiers_for_format(DRMFormat /*format*/) const -> std::vector<uint64_t>
+{
+    // TODO: Pull out off KMS plane info
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/platforms/atomic-kms/server/gbm_display_allocator.cpp:43]: No worse than existing platforms

File: src/platforms/atomic-kms/server/gbm_display_allocator.cpp [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+mga::GBMDisplayAllocator::GBMDisplayAllocator(mir::Fd drm_fd, std::shared_ptr<struct gbm_device> gbm, geom::Size size)
+    : fd{std::move(drm_fd)},
+      gbm{std::move(gbm)},
+      size{size}
+{
+}
+
+auto mga::GBMDisplayAllocator::supported_formats() const -> std::vector<DRMFormat>
+{
+    // TODO: Pull out of KMS plane info
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/platforms/atomic-kms/server/gbm_display_allocator.cpp:37]: No worse than existing platforms

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
This appears to work with one or two monitors (I'm using Miriway/mir-pr3525)

However, when I plug or unplug an external monitor I get:

```
<- ERROR - > atomic-kms: Failed to schedule page flip: Invalid arguement (22)
Fatal signal received
...
``` 

[update]

Built Mir locally, with the "cleanup" code disabled, it didn't crash as consistently, but did after a couple of cycles.

The core wasn't as useful as I hoped: it died in an "out of order" shutdown (which is secondary to the problem causing the shutdown, but also worthy of investigation).

**Comment by AlanGriffiths** [comment]
> This appears to work with one or two monitors (I'm using Miriway/mir-pr3525)
> 
> However, when I plug or unplug an external monitor I get:
> 
> ```
> <- ERROR - > atomic-kms: Failed to schedule page flip: Invalid arguement (22)
> Fatal signal received
> ...
> ```

Rats! This is still happening. (And the warning I introduced above isn't in the log). It must be a different failure mode than the one I fixed above.

**Review by AlanGriffiths — APPROVED** [review]
With the last fix, I'm content. Someone else should cast a vote too (especially about my last commit)

**Comment by Saviq** [comment]
> Someone else should cast a vote too (especially about my last commit)

Looks legit to me, @RAOF feel free to merge if you agree with where the exception is caught.

**Comment by RAOF** [comment]
> AHA! It took writing a script to try and reproduce this, but I now can. (The relevant magic is that you need to trigger the modeset _from the terminal that will be hung_)
> 
> What's happening here is that the final frame submitted by the terminal is racing the modeset sequence.
> 
> When it stops updating, what's happening is the terminal is waiting on a frame event from its last-buffer-before-modeset before submitting any new frames. Mir is not sending a frame event until _something else_ triggers a composition.
> 
> So this is a problem somewhere in our buffer queuing, I think.

Just to loop back to this; I can trivially reproduce this on `gbm-kms`, too.

**Comment by RAOF** [comment]
Here are a couple of extra fixes - [one trivial](https://github.com/canonical/mir/pull/3525/commits/a816adbb49357ece4d5c71ce40d12dac08461507), and one thread race use-after-free.

Finally, I've identified a problem when disconnecting and re-connecting an external display to a different port on my laptop - we're not disassociating the CRTC from the old connector, so we run into resource conflicts. The fix *there* will be to update `clear_connected_unused_outputs` to also clear outputs which are *not* connected, but which have resources set up.

**Comment by RAOF** [comment]
Bah, that doesn't fix it for some reason; I still see a CRTC connected to my disconnected output.

File: src/platforms/atomic-kms/server/kms/display.cpp [hunk 15]
```diff
@@ -262,6 +262,8 @@ void mga::Display::clear_connected_unused_outputs()
 {
     current_display_configuration.for_each_output([&](DisplayConfigurationOutput const& conf_output)
     {
+        auto kms_output = current_display_configuration.get_output_for(conf_output.id);
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/display.cpp:265]:
```suggestion
        auto const kms_output = current_display_configuration.get_output_for(conf_output.id);
```

**Review by AlanGriffiths — APPROVED** [review]
Works well enough to land

**Comment by AlanGriffiths** [comment]
> Works well enough to land

OTOH, I've tried using this a while and found that docking element to the right of my rightmost monitor crashes Miriway.

**Comment by RAOF** [comment]
Now properly supports unplugging a monitor and plugging it in to a different port!

File: src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp [hunk 16]
```diff
@@ -0,0 +1,911 @@
...
+    catch (...)
+    {
+        /*
+         * In order to actually clear the output, we need to have a crtc
+         * connected to the output/connector so that we can disconnect
+         * it. However, not being able to get a crtc is OK, since it means
+         * that the output cannot be displaying anything anyway.
+         */
+        return;
+    }
```
Comment (AlanGriffiths) [inline src/platforms/atomic-kms/server/kms/atomic_kms_output.cpp:385]: A try block around a single line of code followed by a catch block that silently eats everything suggests we are using the wrong reporting mechanism.

**Comment by AlanGriffiths** [comment]
> and will be preferentially used on devices with atomic support

This isn't my experience. I have to explicitly choose `platform-display-libs=mir:atomic-kms` (defaulting to the well tested `mir:gbm-kms` isn't bad, just not what is advertised above)

**Comment by AlanGriffiths** [comment]
> This isn't my experience. I have to explicitly choose `platform-display-libs=mir:atomic-kms` (defaulting to the well tested `mir:gbm-kms` isn't bad, just not what is advertised above)

As that's been confirmed by @Saviq, I'll fix the text

**Review by AlanGriffiths — APPROVED** [review]

**Pull request merged**